src/hooks/test/use-validate-on-submit-test.js

```diff
@@ -0,0 +1,134 @@
+import { mount } from 'enzyme';
+import { useState, useId } from 'preact/hooks';
+
+import { Input } from '../../components/input';
+import { useValidateOnSubmit } from '../use-validate-on-submit';
+
+// Example of a simple form that uses this hook to implement custom display of
+// validation messages.
+function CustomForm({ onSelectURL }) {
+  const [url, setURL] = useState();
+  const [error, setError] = useState();
+  const errorId = useId();
+
+  const onSubmit = useValidateOnSubmit(() => {
+    onSelectURL(url);
+  });
+
+  const onChangeURL = event => {
+    const url = event.target.value;
+    setURL(undefined);
+
+    try {
+      new URL(url);
+      setError(undefined);
+      setURL(url);
+    } catch {
+      setError('Not a valid URL');
+    }
+  };
+
+  return (
+    <form onSubmit={onSubmit} noValidate>
+      <Input
+        onChange={onChangeURL}
+        required
+        aria-label="URL"
+        aria-describedby={errorId}
+        error={error}
+      />
+      <div id={errorId}>{error}</div>
+      <button type="submit">Submit</button>
+    </form>
+  );
+}
+
+describe('useValidateOnSubmit', () => {
+  function changeURL(form, url) {
+    const input = form.find('input');
+    input.getDOMNode().value = url;
+    input.simulate('change');
+  }
+
+  function submitForm(form) {
+    form.find('form').simulate('submit');
+  }
+
+  it('should invoke callback if form has no errors', () => {
+    const onSelectURL = sinon.stub();
+    const form = mount(<CustomForm onSelectURL={onSelectURL} />);
+    changeURL(form, 'https://example.com');
+    submitForm(form);
+    assert.calledWith(onSelectURL, 'https://example.com');
+  });
+
+  it('should invoke change event for empty, required inputs', () => {
+    const onSelectURL = sinon.stub();
+    const form = mount(<CustomForm onSelectURL={onSelectURL} />);
+    const input = form.find('input');
+    const onChange = sinon.stub();
+    input.getDOMNode().addEventListener('change', onChange);
+
+    submitForm(form);
+
+    assert.notCalled(onSelectURL);
+    assert.calledOnce(onChange);
+  });
+
+  it('should not invoke callback if form has errors', () => {
+    const onSelectURL = sinon.stub();
+    const form = mount(<CustomForm onSelectURL={onSelectURL} />);
+    changeURL(form, 'not valid');
+    submitForm(form);
+    assert.notCalled(onSelectURL);
+  });
+
+  it('should focus first input with an error', () => {
+    const onSelectURL = sinon.stub();
+    const form = mount(<CustomForm onSelectURL={onSelectURL} />);
+    changeURL(form, 'not valid');
+
+    const input = form.find('input').getDOMNode();
+    const focusStub = sinon.stub(input, 'focus');
+    submitForm(form);
+
+    assert.calledOnce(focusStub);
```

Comment:
Perhaps you could attach the component to the DOM when mounting it (`mount(..., { attachTo: container })`), and then check if `document.activeElement` is equal to `form.find('input').getDOMNode()` instead of stubbing the focus method. But I have to admit I have faced flaky tests when trying to do that, so I'm fine with this if the other approach is problematic.

Reply:
> But I have to admit I have faced flaky tests when trying to do that, so I'm fine with this if the other approach is problematic.

Indeed that's the rationale here - a call to `HTMLElement.focus` is much easier to check for as there is no asynchrony involved.